benchmarks/src/main/java/io/grpc/benchmarks/qps/QpsAsyncClient.java

```diff
@@ -0,0 +1,411 @@
+
+import grpc.testing.Qpstest.Payload;
+import grpc.testing.TestServiceGrpc;
+import io.grpc.Channel;
+import io.grpc.ChannelImpl;
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+import io.grpc.transport.netty.NegotiationType;
+import io.grpc.transport.netty.NettyChannelBuilder;
+import io.grpc.transport.okhttp.OkHttpChannelBuilder;
+import io.netty.handler.ssl.SslContext;
+import org.HdrHistogram.Histogram;
+import org.HdrHistogram.HistogramIterationValue;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class QpsAsyncClient {
+  // The histogram can record values between 1 microsecond and 1 min.
+  private static final long HISTOGRAM_MAX_VALUE = 60000000L;
+  // Value quantization will be no larger than 1/10^3 = 0.1%.
+  private static final int HISTOGRAM_PRECISION = 3;
+
+  private final ClientConfiguration config;
+
+  public QpsAsyncClient(String[] args) {
+    config = new ClientConfiguration(args);
+  }
+
+  public void run() throws Exception {
+    SimpleRequest req = newRequest();
+
+    List<Channel> channels = new ArrayList<Channel>(config.channels);
+    for (int i = 0; i < config.channels; i++) {
+      channels.add(newChannel());
+    }
+
+    // Do a warmup first. It's the same as the actual benchmark, except that
+    // we ignore the statistics.
+    warmup(req, channels);
+
+    long startTime = System.nanoTime();
+    long endTime = startTime + TimeUnit.SECONDS.toNanos(config.duration);
+    List<Histogram> histograms = doBenchmark(req, channels, endTime);
+    long elapsedTime = System.nanoTime() - startTime;
+
+    Histogram merged = merge(histograms);
+
+    printStats(merged, elapsedTime);
+    dumpHistogram(merged);
+    shutdown(channels);
+  }
+
+  private SimpleRequest newRequest() {
+    ByteString body = ByteString.copyFrom(new byte[config.clientPayload]);
+    Payload payload = Payload.newBuilder().setType(config.payloadType).setBody(body).build();
+
+    return SimpleRequest.newBuilder()
+            .setResponseType(config.payloadType)
+            .setResponseSize(config.serverPayload)
+            .setPayload(payload)
+            .build();
+  }
+
+  private void warmup(SimpleRequest req, List<Channel> channels) throws Exception {
+    long endTime = System.nanoTime() + TimeUnit.SECONDS.toNanos(config.warmupDuration);
+    doBenchmark(req, channels, endTime);
+    // I don't know if this helps, but it doesn't hurt trying. We sometimes run warmups
+    // of several minutes at full load and it would be nice to start the actual benchmark
+    // with a clean heap.
+    System.gc();
```

Comment:
@nmittler that's generally not how it works. HotSpot has it's own compiler threads and swaps methods out with their compiled counterparts on the fly. We have no information if and how long to sleep for, also instead of sleeping one could just increase the warmup duration if he feels / knows that some optimization takes longer to kick in.